File: src/main/java/no/stunor/origo/eventorapi/services/converter/PersonConverter.java
package no.stunor.origo.eventorapi.services.converter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.iof.eventor.Role;

import no.stunor.origo.eventorapi.model.firestore.Eventor;
import no.stunor.origo.eventorapi.model.firestore.Person;
import no.stunor.origo.eventorapi.model.origo.person.CompetitorPerson;
import no.stunor.origo.eventorapi.model.origo.person.PersonName;

public class PersonConverter {

    public static Person convertPerson(org.iof.eventor.Person eventorPerson, Eventor eventor) {
        return new Person(
            null,
            eventor.getEventorId(),
            eventorPerson.getPersonId().getContent(),
            convertPersonName(eventorPerson.getPersonName()),
            eventorPerson.getBirthDate()!= null ? Integer.parseInt(eventorPerson.getBirthDate().getDate().getContent().substring(0,4)) : null,
            eventorPerson.getNationality() != null ? eventorPerson.getNationality().getCountry().getAlpha3().getValue() : null,
            convertGender(eventorPerson.getSex()),
            new ArrayList<>(),
            convertMemberships(eventorPerson.getRole()),
            ContactConverter.convertPhone(eventorPerson.getTele()),
            ContactConverter.convertEmail(eventorPerson.getTele())
        );
    }

    public static CompetitorPerson convertCompetitor(org.iof.eventor.Person person, Eventor eventor) {
        return new CompetitorPerson(
            eventor.getEventorId(),
            person.getPersonId().getContent(),
            convertPersonName(person.getPersonName()),
            person.getBirthDate()!= null ? Integer.parseInt(person.getBirthDate().getDate().getContent().substring(0,4)) : null,
            person.getNationality() != null && person.getNationality().getCountry() != null ? person.getNationality().getCountry().getAlpha3().getValue() : null,
            convertGender(person.getSex())
        );
    }

    private static String convertGender(String sex) {
        if(sex == null){
            return "OTHER";
        }
        switch(sex){
            case "M" : return "MAN";
            case "F" : return "WOMAN";
            default  : return "OTHER";
        }
    }

    public static PersonName convertPersonName(org.iof.eventor.PersonName personName) {
        StringBuilder given = new StringBuilder();
        for(int i = 0; i < personName.getGiven().size(); i++) {
            for (int j = 1; j <= personName.getGiven().size(); j++) {
                if(Integer.parseInt(personName.getGiven().get(i).getSequence()) == j){
                    if(!given.toString().equals("")){
                        given.append(" ");
                    }
                    given.append(personName.getGiven().get(i).getContent());
                }
            }
        }

        return new PersonName(personName.getFamily().getContent(), given.toString());
    }
    private static Map<String, String> convertMemberships(List<Role> roles) {
        Map<String, Integer> highestRole= new HashMap<>();

        for (Role role : roles){
            if(role.getRoleTypeId().getContent().equals("2")){
                role.getRoleTypeId().setContent("10");;
            }

            if(!highestRole.containsKey(role.getOrganisationId().getContent())){
                highestRole.put(role.getOrganisationId().getContent(), Integer.parseInt(role.getRoleTypeId().getContent()));
            } else if(Integer.parseInt(role.getRoleTypeId().getContent()) > highestRole.get(role.getOrganisationId().getContent())){
                highestRole.put(role.getOrganisationId().getContent(), Integer.parseInt(role.getRoleTypeId().getContent()));
            }
        }

        Map<String, String> memberships = new HashMap<>();

        for(String orgId : highestRole.keySet().stream().toList()){
            if(highestRole.get(orgId) == 1){
                memberships.put(orgId, "MEMBER");
            } else if (highestRole.get(orgId) == 3 || highestRole.get(orgId) == 5) {
                memberships.put(orgId, "ORGANISER");
            } else if (highestRole.get(orgId) == 10) {
                memberships.put(orgId, "ADMIN");
            }
        }


        return memberships;
    }
    
}

File: src/main/java/no/stunor/origo/eventorapi/services/converter/PersonConverter.kt
package no.stunor.origo.eventorapi.services.converter

import no.stunor.origo.eventorapi.model.Eventor
import no.stunor.origo.eventorapi.model.origo.CompetitorPerson
import no.stunor.origo.eventorapi.model.person.Gender
import no.stunor.origo.eventorapi.model.person.MembershipType
import no.stunor.origo.eventorapi.model.person.Person
import no.stunor.origo.eventorapi.model.person.PersonName
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class PersonConverter {

    @Autowired
    private lateinit var contactConverter: ContactConverter
    fun convertPerson(eventorPerson: org.iof.eventor.Person, eventor: Eventor): Person {
        return Person(
                null,
                eventor.eventorId,
                eventorPerson.personId.content,
                convertPersonName(eventorPerson.personName),
                eventorPerson.birthDate.date.content.substring(0, 4).toInt(),
                eventorPerson.nationality.country.alpha3.value,
                convertGender(eventorPerson.sex),
                ArrayList(),
                contactConverter.convertPhone(eventorPerson.tele),
                contactConverter.convertEmail(eventorPerson.tele),
                convertMemberships(eventorPerson.role)
        )
    }

    fun convertCompetitor(person: org.iof.eventor.Person, eventor: Eventor): CompetitorPerson {
        return CompetitorPerson(
                eventor.eventorId,
                person.personId.content,
                convertPersonName(person.personName),
                person.birthDate.date.content.substring(0, 4).toInt(),
                person.nationality.country.alpha3.value,
                convertGender(person.sex)
        )
    }

    fun convertBirthYear(birthDate: org.iof.eventor.BirthDate?): Int? {
        if (birthDate == null) {
            return null
        }
        return birthDate.date.content.substring(0, 4).toInt()
    }
    fun convertGender(sex: String?): Gender {
        if (sex == null) {
            return Gender.OTHER
        }
        return when (sex) {
            "M" -> Gender.MAN
            "F" -> Gender.WOMAN
            else -> Gender.OTHER
        }
    }

    fun convertPersonName(personName: org.iof.eventor.PersonName): PersonName {
        val given = StringBuilder()
        for (i in personName.given.indices) {
            for (j in 1..personName.given.size) {
                if (personName.given[i].sequence.toInt() == j) {
                    if (given.toString() != "") {
                        given.append(" ")
                    }
                    given.append(personName.given[i].content)
                }
            }
        }
        return PersonName(personName.family.content, given.toString())
    }

    private fun convertMemberships(roles: List<org.iof.eventor.Role>): Map<String, MembershipType> {
        val highestRole: MutableMap<String, Int> = HashMap()

        for (role in roles) {
            if (role.roleTypeId.content == "2") {
                role.roleTypeId.content = "10"
            }

            if (!highestRole.containsKey(role.organisationId.content)) {
                highestRole[role.organisationId.content] = role.roleTypeId.content.toInt()
            } else if (role.roleTypeId.content.toInt() > highestRole[role.organisationId.content]!!) {
                highestRole[role.organisationId.content] = role.roleTypeId.content.toInt()
            }
        }

        val memberships: MutableMap<String, MembershipType> = HashMap()

        for (orgId in highestRole.keys.stream().toList()) {
            if (highestRole[orgId] == 1) {
                memberships[orgId] = MembershipType.MEMBER
            } else if (highestRole[orgId] == 3 || highestRole[orgId] == 5) {
                memberships[orgId] = MembershipType.ORGANISER
            } else if (highestRole[orgId] == 10) {
                memberships[orgId] = MembershipType.ADMIN
            }
        }


        return memberships
    }
}
